src/ophyd_async/core/signal.py

```diff
@@ -73,19 +76,49 @@ async def connect(
         backend: Optional[SignalBackend[T]] = None,
     ):
         if backend:
+            if self._backend and backend is not self._backend:
+                # test if the backend from a previous connection is the same
+                # reject even if there was no initial backend
+                raise ValueError(
+                    "Backend at connection different from initialised one."
+                )
+
             if self._initial_backend and backend is not self._initial_backend:
                 raise ValueError(
                     "Backend at connection different from initialised one."
                 )
             self._backend = backend
+        if (
+            self._previous_connect_was_mock is not None
+            and self._previous_connect_was_mock != mock
```

Comment:
You never set `self._previous_connect_was_mock = mock` after this check.

Reply:
fixed